### PR TITLE
[MOBL-116] Customisable action buttons for push notifications.

### PR DESCRIPTION
**Pull request by rahulrvp**
[MOBL-116] Customisable action buttons for push notifications.

[MOBL-116]: https://blueshift.atlassian.net/browse/MOBL-116

### CODE REVIEW

File: android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java [hunk 1]
```diff
@@ -578,20 +583,49 @@ public static void invokePushDelivered(Context context, Message message) {
...
 
             if (Blueshift.getBlueshiftPushListener() != null) {
                 Blueshift.getBlueshiftPushListener().onPushClicked(attr);
             }
         }
 
-        HashMap<String, Object> clickAttr = new HashMap<>();
-        if (clickUrl != null) clickAttr.put(BlueshiftConstants.KEY_CLICK_URL, clickUrl);
+        HashMap<String, Object> clickAttr = extras != null ? new HashMap<>(extras) : null;
         Blueshift.getInstance(context).trackNotificationClick(message, clickAttr);
```
Comment (ketanshikharebsft) [inline android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java:599]: Is it okay here if `clickAttr` parameter is nil?

Comment (ketanshikharebsft) [inline android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java:599]: Is it okay here if clickAttr parameter is nil?

File: android-sdk/src/main/java/com/blueshift/Blueshift.java [hunk 2]
```diff
@@ -1379,6 +1379,9 @@ private boolean sendNotificationEvent(String action, HashMap<String, Object> cam
 
             String paramsUrl = q.toString();
             if (!TextUtils.isEmpty(paramsUrl)) {
+                // replace whitespace with %20 to avoid URL damage.
+                paramsUrl = paramsUrl.replace(" ", "%20");
```
Comment (ketanshikharebsft) [inline android-sdk/src/main/java/com/blueshift/Blueshift.java:1383]: Can we use constant for %20 ?

File: android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java [hunk 3]
```diff
@@ -578,20 +583,49 @@ public static void invokePushDelivered(Context context, Message message) {
...
+    public static List<NotificationCompat.Action> getActions(Context context, Message message, int notificationId) {
+        List<NotificationCompat.Action> actionList = null;
+        if (context != null && message != null && message.hasActions()) {
+            actionList = new ArrayList<>();
+
+            List<Action> actionItems = message.getActions();
+            if (actionItems != null) {
+                // Android supports max 3 actions on a notification.
+                // https://developer.android.com/training/notify-user/build-notification#Actions
+                int count = Math.min(actionItems.size(), 3);
```
Comment (ketanshikharebsft) [inline android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java:611]: Can we use a constant for max actions value?

**Review by ketanshikharebsft — APPROVED** [review]
LGTM

File: android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java (hunk 1 above)
Comment (rahulrvp) [reply android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java:599]: yes, `trackNotificationClick` can accept `null` as `clickAttr`

File: android-sdk/src/main/java/com/blueshift/Blueshift.java (hunk 2 above)
Comment (rahulrvp) [reply android-sdk/src/main/java/com/blueshift/Blueshift.java:1383]: sure, will do that

File: android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java (hunk 3 above)
Comment (rahulrvp) [reply android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java:611]: sure

Comment (rahulrvp) [reply android-sdk/src/main/java/com/blueshift/util/NotificationUtils.java:611]: I changed my mind. Should we use memory to hold this variable? This is only used here and nowhere else.

**Pull request merged**